File: Python/linked_list/delete_in_DLL.py
"""
DATE: April 5, 2026
PATTERN: Doubly Linked List, Deletion at Tail

APPROACHES:
1. BRUTE FORCE:
- Step 1: Copy list without last node into new list.
- TIME: O(n) | SPACE: O(n)

2. BETTER:
- Step 1: Find length, walk to node before last, null next.
- TIME: O(n) | SPACE: O(1)

3. OPTIMAL:
- Step 1: Walk to tail; set prev.next=None and clear back of removed node.
- TIME: O(n) | SPACE: O(1)

AHA-MOMENT: Tail delete in DLL is O(n) to reach tail—four pointer updates once at tail.
"""

class Node:
    def __init__(self,data1,next1=None,back1=None):
        self.data=data1
        self.next=next1
        self.back=back1

class Solution:
    def convert_arr_2DLL(self,arr):
        head=Node(arr[0])
        prev=head
        for i in range(1,len(arr)):
            temp=Node(arr[i],None,prev)
            prev.next=temp
            prev=temp

        return head

    def delete_in_DLL(self,head):
        if not head:
            return None

        if not head.next:
            return None

        temp=head
        while temp.next:
            temp=temp.next

        temp.back.next = None
        temp.back = None



    def printLL(self,head):
        temp=head
        while temp:
            print(temp.data, end=" ")
            temp=temp.next
        
        print()

if __name__=="__main__":
    sol=Solution()
    arr=list((1,3,4,5,6,7,3))
    head=sol.convert_arr_2DLL(arr)

    print("DLL Before Deleting:")
    sol.printLL(head)
    
    sol.delete_in_DLL(head)
    print("DLL after Deleting:")
    sol.printLL(head)
        

File: CPP/linked_list/delete_in_DLL.cpp
/*
 * DATE: April 5, 2026
 * PATTERN: Doubly Linked List, Pointer Surgery
 *
 * APPROACHES:
 * 1. BRUTE FORCE:
 * - Step 1: Copy list to vector, remove element by index, rebuild DLL from vector.
 * - TIME: O(n) | SPACE: O(n)
 *
 * 2. BETTER:
 * - Step 1: Find node with value/key via linear search, relink prev/next for that node only.
 * - TIME: O(n) search | SPACE: O(1)
 *
 * 3. OPTIMAL (delete tail as in code):
 * - Step 1: Walk to last node; set prev->next=nullptr, clear back of removed node.
 * - TIME: O(n) to reach tail | SPACE: O(1)
 *
 * AHA-MOMENT: In a DLL, deletion is local: fix two neighbors’ pointers—no full list copy.
 */

#include<bits/stdc++.h>
using namespace std;

class Node
{
    public:
    int val;
    Node * next;
    Node * back;

    Node(int val1,Node* next1,Node* back1)
    {
        val=val1;
        next=next1;
        back=back1;

    }
    Node(int val1)
    {
        val=val1;
        next=nullptr;
        back=nullptr;
    }
};

class Solution
{
    public:
    Node* convert_arr_2DLL(vector<int>& arr)
    {
        Node * head=new Node(arr[0]);
        Node * prev=head;

        for(int i=1;i<arr.size();i++)
        {
            Node * temp=new Node(arr[i],nullptr,prev);
            prev->next=temp;
            prev=temp;
        }
        return head;
    }

    Node * delete_in_2DLL(Node * head)
    {
        Node * temp=head;
        while(temp->next!=nullptr)
        {
            temp=temp->next;
        }
        temp->back->next=nullptr;
        temp->back=nullptr;
        return head;
    }

    void printList(Node* head)
    {
        Node* temp=head;
        while(temp!=nullptr)
        {
            cout<<temp->val<<" ";
            temp=temp->next;
        }
    }   
    
};

int main()
{
    Solution sol;
    vector<int> arr={12,5,8,7,4};
    Node* head=sol.convert_arr_2DLL(arr);

    cout<<"Doubly Linked List Initially:"<<endl;

    sol.printList(head);

    head=sol.delete_in_2DLL(head);

    cout<<"Doubly Linked List After Inserting at tail:"<<endl;
    sol.printList(head);

    return 0;


}